anyio/__init__.py

```diff
@@ -292,34 +299,65 @@ def aopen(file: Union[str, 'os.PathLike', int], mode: str = 'r', buffering: int
         :exc:`~ssl.SSLEOFError` may be raised during reads from the stream.
         Some protocols, such as HTTP, require this option to be ``False``.
         See :meth:`~ssl.SSLContext.wrap_socket` for details.
+    :param happy_eyeballs_delay: delay (in seconds) before starting the next connection attempt
     :return: a socket stream object
+    :raises OSError: if the connection attempt fails
 
     """
+    # Placed here due to https://github.com/python/mypy/issues/7057
+    stream = None  # type: Optional[SocketStream]
+
+    async def try_connect(af: int, addr: str, delay: float):
+        nonlocal stream
+
+        if delay:
+            await sleep(delay)
+
+        raw_socket = socket.socket(af, socket.SOCK_STREAM)
+        sock = asynclib.Socket(raw_socket)
+        try:
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+            if interface is not None and bind_port is not None:
+                await sock.bind((interface, bind_port))
+
+            await sock.connect((addr, port))
+        except OSError as exc:
+            oserrors.append(exc)
+            await sock.close()
+            return
+        except BaseException:
+            await sock.close()
+            raise
+
+        assert stream is None
+        stream = _networking.SocketStream(sock, ssl_context, str(address), tls_standard_compatible)
+        await tg.cancel_scope.cancel()
+
+    asynclib = _get_asynclib()
     interface, family = None, 0  # type: Optional[str], int
     if bind_host:
         interface, family, _v6only = await _networking.get_bind_address(bind_host)
 
     # getaddrinfo() will raise an exception if name resolution fails
-    address = str(address)
-    addrlist = await run_in_thread(socket.getaddrinfo, address, port, family, socket.SOCK_STREAM)
-    family, type_, proto, _cn, sa = addrlist[0]
-    raw_socket = socket.socket(family, type_, proto)
-    sock = _get_asynclib().Socket(raw_socket)
-    try:
-        sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
-        if interface is not None and bind_port is not None:
-            await sock.bind((interface, bind_port))
+    addrlist = await run_in_thread(socket.getaddrinfo, str(address), port, family,
+                                   socket.SOCK_STREAM)
 
-        await sock.connect(sa)
-        stream = _networking.SocketStream(sock, ssl_context, address, tls_standard_compatible)
+    # Sort the list so that IPv4 addresses are tried last
+    addresses = sorted(((item[0], item[-1][0]) for item in addrlist),
```

Comment:
Isn't this not optimal in the case that IPv6 isn't reachable and IPv4 is? The RFC recommends the first two attempts have different families.

Reply:
I'll go take a closer look at the RFC and get back to you on this.

Reply:
Yeah, my interpretation of the RFC text is "try the first AAAA address first, if that fails, then the first A address, then the rest in an unspecified order". I'll sort the addresses accordingly.

Reply:
My latest commits should address this.